exist-core/src/main/java/org/exist/xquery/functions/securitymanager/IdFunction.java

```diff
@@ -77,13 +77,18 @@ private org.exist.dom.memtree.DocumentImpl functionId() {
 
             builder.startElement(new QName("id", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
 
-            builder.startElement(new QName("real", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
-            subjectToXml(builder, context.getRealUser());
-            builder.endElement();
+            final Subject realUser = context.getRealUser();
+            if (realUser != null) {
+                builder.startElement(new QName("real", SecurityManagerModule.NAMESPACE_URI, SecurityManagerModule.PREFIX), null);
+                subjectToXml(builder, realUser);
+                builder.endElement();
+            }
 
-            if (!sameUserWithSameGroups(context.getRealUser(), context.getEffectiveUser())) {
+            final Subject effectiveUser = context.getEffectiveUser();
+            if (effectiveUser != null && (
+                    realUser == null || !sameUserWithSameGroups(realUser, effectiveUser))) {
```

Comment:
`realUser` will never be `null`

Reply:
If context.realUser returns null - then that indicates a serious bug elsewhere that needs to be solved. Changing the code here will only hide that bug and make it much harder to find!

Reply:
throw XPathexception instead ?

Reply:
I would consider actually throwing a RuntimeException and shutting down - as at this point the security of the system is completely compromised

Reply:
I'll try that and see if tests still pass